### PR TITLE
Do not cancel ongoing recovery for noop copy on broken node

### LINKED ISSUE #47974: testDoNotInfinitelyWaitForMapping fails
This test starts failing since #46959 where we cancel an ongoing recovery if we find a new copy that can perform a noop recovery. 

> [2019-10-13T06:29:10,208][WARN ][o.e.c.r.a.AllocationService] [node_t0] failing shard [failed shard, shard [test][0], node[d1YsSCuCScGq6Micn4jKqQ], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=o1Qiiu9PSO-fB6M84wsdtA], unassigned_info[[reason=REALLOCATED_REPLICA], at[2019-10-13T17:29:10.107Z], delayed=false, details[existing allocation of replica to [{node_t1}{Z7bFtb4bSyWxw8hW8jhLjQ}{ApVGNCrUQkyIG-5hbQuRiw}{127.0.0.1}{127.0.0.1:44756}{dim}] cancelled, can perform a noop recovery on [{node_t2}{d1YsSCuCScGq6Micn4jKqQ}{vJfgfJ52RwqCdjqnUdYOAQ}{127.0.0.1}{127.0.0.1:44089}{dim}]], allocation_status[no_attempt]], message [shard failure, reason [index id[u0] origin[PEER_RECOVERY] seq#[0]]], markAsStale [true], failure [org.elasticsearch.index.mapper.MapperParsingException: simulate mapping parsing error

CI: https://gradle-enterprise.elastic.co/s/zbewn2l6ksvd2/tests/kyv2y2z3r4v7m-bgzwhe6nv7k4c

Relates #46959

### PR DESCRIPTION
**Pull request by dnhatn**
Today the replica allocator can repeatedly cancel an ongoing recovery for a copy on a broken node if that copy can perform a noop recovery. This loop can be happen endlessly (see [testDoNotInfinitelyWaitForMapping](https://gradle-enterprise.elastic.co/s/zbewn2l6ksvd2/tests/kyv2y2z3r4v7m-bgzwhe6nv7k4c)). We should detect and avoid canceling in this situation.

Closes #47974

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java [hunk 1]
```diff
@@ -282,6 +285,39 @@ public void testPreferCopyWithHighestMatchingOperations() throws Exception {
         assertThat(internalCluster().nodesInclude(indexName), allOf(hasItem(nodeWithHigherMatching), not(hasItem(nodeWithLowerMatching))));
     }
 
+    /**
+     * Make sure that we do not repeatedly cancel an ongoing recovery for a noop copy on a broken node.
+     */
+    public void testDoNotCancelRecoveryForBrokenNode() throws Exception {
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java]: Just a drive by comment from trying understand this one a little :)

It seems this test isn't really reproducing the issue much better than `testDoNotInfinitelyWaitForMapping` (only fails randomly at a low rate). Could we make this one more deterministic?

**Review by DaveCTurner — COMMENTED** [review]
My concern with using a simple boolean is that it will suppress no-op recoveries to other nodes that might succeed. I think we should track _which_ node failed the no-op allocation.

What would happen if instead we kept track of all the nodes on which this shard failed during initialization (whether no-op or otherwise) and ignored them all in the `ReplicaShardAllocator`? I am thinking particularly of https://github.com/elastic/elasticsearch/issues/18417 - with a list of past failures we could also prefer to avoid those nodes in the `BalancedShardsAllocator`.

**Comment by dnhatn** [comment]
@DaveCTurner Thanks for looking. We would make better decisions with a list of failed nodes. We need to make sure that the list is bounded. I used a boolean to avoid adding more load to the cluster state.

File: server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java (hunk 1 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java]: ++. I adjusted this test in https://github.com/elastic/elasticsearch/pull/48265/commits/600c3dee8905b53243d9c12983b365fbbecf151e.

**Comment by DaveCTurner** [comment]
I think a list of nodes (or node IDs) would not be too much to add to the cluster state and would naturally be bounded: we would only add to it on an `ALLOCATION_FAILED` which only happens a few times thanks to the `MaxRetryAllocationDecider`; the list should of course be cleared in `resetFailedAllocationCounter()`.

**Comment by dnhatn** [comment]
@DaveCTurner Good point. I will apply your feedback. Thank you.

File: server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java [hunk 2]
```diff
@@ -110,7 +111,7 @@ public void processExistingRecoveries(RoutingAllocation allocation) {
                             "existing allocation of replica to [" + currentNode + "] cancelled, can perform a noop recovery on ["+
                                 nodeWithHighestMatch + "]",
                             null, 0, allocation.getCurrentNanoTime(), System.currentTimeMillis(), false,
-                            UnassignedInfo.AllocationStatus.NO_ATTEMPT);
+                            UnassignedInfo.AllocationStatus.NO_ATTEMPT, Collections.emptySet());
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java]: I am not sure if we should carry over the failedNodeIds and failedCount here. If we have two broken nodes with noop copies, then we can repeatedly cancel an ongoing recovery for them. Note that this is not an issue with the current implementation as we ignore the failed nodes for the next round of allocation only.

**Comment by dnhatn** [comment]
@DaveCTurner It's ready again. Can you please take another look? Thank you.

**Review by DaveCTurner — COMMENTED** [review]
Sorry for the delayed review @dnhatn, I thought I submitted this a few days ago but apparently not.

File: server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java [hunk 3]
```diff
@@ -245,10 +251,12 @@ public UnassignedInfo(Reason reason, @Nullable String message, @Nullable Excepti
         this.failure = failure;
         this.failedAllocations = failedAllocations;
         this.lastAllocationStatus = Objects.requireNonNull(lastAllocationStatus);
+        this.failedNodeIds = Collections.unmodifiableSet(failedNodeIds);
         assert (failedAllocations > 0) == (reason == Reason.ALLOCATION_FAILED) :
             "failedAllocations: " + failedAllocations + " for reason " + reason;
         assert !(message == null && failure != null) : "provide a message if a failure exception is provided";
         assert !(delayed && reason != Reason.NODE_LEFT) : "shard can only be delayed if it is unassigned due to a node leaving";
+        assert failedAllocations >= failedNodeIds.size() : "failedAllocations: " + failedAllocations + " failedNodeIds: " + failedNodeIds;
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java]: Unfortunately this does not hold when we start carrying over the failedNodeIds in `processExistingRecoveries` (unless we also carry over the failedAllocations, which is reset there).

**Review by henningandersen — COMMENTED** [review]
Thanks for working on this @dnhatn, I left a few comments to consider.

File: server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java [hunk 4]
```diff
@@ -213,6 +216,7 @@ public String value() {
     private final String message;
     private final Exception failure;
     private final int failedAllocations;
+    private final Set<String> failedNodeIds;
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java]: ++. I added in https://github.com/elastic/elasticsearch/pull/48265/commits/05e7691cd18d45a87436821ee7817907f3c7b5d6.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java [hunk 5]
```diff
@@ -194,10 +196,18 @@ public ClusterState applyFailedShards(final ClusterState clusterState, final Lis
                         shardToFail.shardId(), shardToFail, failedShard);
                 }
                 int failedAllocations = failedShard.unassignedInfo() != null ? failedShard.unassignedInfo().getNumFailedAllocations() : 0;
+                final Set<String> failedNodeIds;
+                if (failedShard.unassignedInfo() != null) {
+                    failedNodeIds = new HashSet<>(failedShard.unassignedInfo().getFailedNodeIds().size() + 1);
+                    failedNodeIds.addAll(failedShard.unassignedInfo().getFailedNodeIds());
+                    failedNodeIds.add(failedShard.currentNodeId());
+                } else {
+                    failedNodeIds = Set.of(failedShard.currentNodeId());
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: Yes, I've adjusted this in https://github.com/elastic/elasticsearch/pull/48265/commits/0dd5cc8c2c59b6c440890b17f5a44b7dac9a4439.

File: server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java [hunk 6]
```diff
@@ -297,14 +299,18 @@ public AllocateUnassignedDecision makeAllocationDecision(final ShardRouting unas
...
-    private MatchingNodes findMatchingNodes(ShardRouting shard, RoutingAllocation allocation,
+    private MatchingNodes findMatchingNodes(ShardRouting shard, RoutingAllocation allocation, boolean ignorePreviousFailedNodes,
                                             DiscoveryNode primaryNode, TransportNodesListShardStoreMetaData.StoreFilesMetaData primaryStore,
                                             AsyncShardFetch.FetchResult<NodeStoreFilesMetaData> data,
                                             boolean explain) {
         Map<DiscoveryNode, MatchingNode> matchingNodes = new HashMap<>();
         Map<String, NodeAllocationResult> nodeDecisions = explain ? new HashMap<>() : null;
         for (Map.Entry<DiscoveryNode, NodeStoreFilesMetaData> nodeStoreEntry : data.getData().entrySet()) {
             DiscoveryNode discoNode = nodeStoreEntry.getKey();
+            if (ignorePreviousFailedNodes
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java]: changed in https://github.com/elastic/elasticsearch/pull/48265/commits/a77b19d2add2dda590a363e000488261f73a3495.

File: server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java [hunk 7]
```diff
@@ -347,6 +363,13 @@ public AllocationStatus getLastAllocationStatus() {
         return lastAllocationStatus;
     }
 
+    /**
+     * Returns a set of nodeId that previously failed to allocate this shard
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java]: ++. Please see https://github.com/elastic/elasticsearch/pull/48265/commits/0dd5cc8c2c59b6c440890b17f5a44b7dac9a4439.

File: server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java [hunk 8]
```diff
@@ -282,6 +284,39 @@ public void testPreferCopyWithHighestMatchingOperations() throws Exception {
...
+                if (brokenNode.equals(connection.getNode().getName())) {
+                    throw new CircuitBreakingException("not enough memory for indexing", 100, 50, CircuitBreaker.Durability.TRANSIENT);
+                }
+            }
+            connection.sendRequest(requestId, action, request, options);
+        });
+        assertAcked(client().admin().indices().prepareUpdateSettings(indexName)
+            .setSettings(Settings.builder().put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 1)));
+        internalCluster().startDataOnlyNode();
+        client().admin().cluster().prepareReroute().setRetryFailed(true).get();
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java]: fixed in https://github.com/elastic/elasticsearch/pull/48265/commits/e27bafe6f1ce02203b9c08dc6608a99d822219d8.

File: server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java (hunk 3 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java]: I pushed https://github.com/elastic/elasticsearch/pull/48265/commits/89c675346b2d932bc07d4f94932766e52c57e2e6.

File: server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java]: I pushed https://github.com/elastic/elasticsearch/pull/48265/commits/89c675346b2d932bc07d4f94932766e52c57e2e6 to carry over the failedNodeIds but not the failedAllocations.

File: server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java]: I have to remove this assertion as we carry over the failedNodeIds but not failedAllocations when canceling a recovery.

**Comment by dnhatn** [comment]
@DaveCTurner @henningandersen Thanks so much for reviewing. This is ready again.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: Hmm. I do agree that if a shard is failed while started we should not count it as an allocation failure (and therefore we should consider it for a future no-op recovery). However I think we should attempt to track all allocation failures here, not just those resulting from `REALLOCATED_REPLICA`. Can we revert 0dd5cc8 except for the change to the `else` branch that should read `failedNodeIds = Collections.emptySet();`?

File: server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java]: I think it's right to clear the failed node IDs when rerouting with `?retry_failed`. However I don't think we need the `?retry_failed` in this test.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java [hunk 9]
```diff
@@ -194,10 +196,19 @@ public ClusterState applyFailedShards(final ClusterState clusterState, final Lis
                         shardToFail.shardId(), shardToFail, failedShard);
                 }
                 int failedAllocations = failedShard.unassignedInfo() != null ? failedShard.unassignedInfo().getNumFailedAllocations() : 0;
+                final Set<String> failedNodeIds;
+                if (failedShard.unassignedInfo() != null &&
+                    failedShard.unassignedInfo().getReason() == UnassignedInfo.Reason.REALLOCATED_REPLICA) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: Duplicating https://github.com/elastic/elasticsearch/pull/48265#discussion_r340525608:

Hmm. I do agree that if a shard is failed while started we should not count it as an allocation failure (and therefore we should consider it for a future no-op recovery). However I think we should attempt to track all allocation failures here, not just those resulting from `REALLOCATED_REPLICA`. Can we revert 0dd5cc8 except for the change to the `else` branch that should read `failedNodeIds = Collections.emptySet();`?

**Review by DaveCTurner — COMMENTED** [review]
I left a couple of responses to threads in an earlier review and duplicated them here.

**Comment by dnhatn** [comment]
@DaveCTurner I've addressed your feedback. Would you mind taking another look? Thank you.

**Comment by gwbrown** [comment]
@dnhatn The test failure in ci/2 here is a known issue (https://github.com/elastic/elasticsearch/issues/48531), feel free to ignore it (it's unmuted to collect logs for troubleshooting purposes).

**Comment by dnhatn** [comment]
Thanks @gwbrown.

@elasticmachine run elasticsearch-ci/2

**Review by DaveCTurner — APPROVED** [review]
Thanks @dnhatn this LGTM

**Comment by dnhatn** [comment]
@DaveCTurner @henningandersen @original-brownbear Thanks for reviewing :).

**Pull request merged**